File: 33-graph-algorithms/14-dijkstra-algorithm-shortest-path.py
import sys

class Graph:
    def __init__(self):
        self.matrix = {} # {'Delhi':[('Mumbai', 4), ('Punjab', 5)]}

    def add_edge(self, u, v, dist, bidirection=True):
        if self.matrix.get(u):
            self.matrix[u].append((v, dist))
        else:
             self.matrix[u] = [(v, dist)]
        if bidirection:
            if self.matrix.get(v):
                self.matrix[v].append((u, dist))
            else:
                self.matrix[v] = [(u, dist)]
        
    
    def print_adj_matrix(self):
        for j in self.matrix:
            print(j, end="->")
            for l in self.matrix.get(j):
                print("( {}, {} )".format(l[0],l[1]), end=" ")
            print()
    
    def dijkstra(self, src, n):
        dist = dict.fromkeys(range(1,n+1),sys.maxsize) # {Delhi:4, ..}
        # for ele in self.matrix:
        #     dist[ele] = sys.maxsize # INF distance
        visited_set = set() # ((4, Delhi), ...)
        dist[src] = 0

        visited_set.add((dist[src],src))

        while len(visited_set):
            iterator = iter(visited_set)
            node = next(iterator) # Delhi
            node_distance = node[0]
            node_name = node[1]
            visited_set.remove(node)

            for child in self.matrix[node_name]:
                if node_distance + child[1] < dist[child[0]]:
                    destination = child[0]
                    to_be_removed = (dist[destination], destination)
                    if to_be_removed in visited_set:
                        visited_set.remove(to_be_removed)
                    dist[destination] = node_distance + child[1] # min. distance
                    visited_set.add((dist[destination], destination))

        # Print distances
        del dist[src]
        for i in dist:
            if dist[i]==sys.maxsize:
                print(-1, end=" ")
            else:
                print(dist[i], end=" ")
        print()







# Main 
t = int(input())
for _ in range(t):
	g = Graph()
	nodes, edges = list(map(int, input().split(' ')))
	for _ in range(edges):
		s,e,d = list(map(int, input().split(' ')))
		g.add_edge(s,e,d)
		
	src = int(input())
	g.dijkstra(src, nodes)

# g = Graph()

# g.add_edge("Amritsar", "Delhi", 1)
# g.add_edge("Amritsar", "Jaipur", 4)
# g.add_edge("Jaipur", "Delhi", 2)
# g.add_edge("Jaipur", "Mumbai", 8)
# g.add_edge("Bhopal", "Agra", 2)
# g.add_edge("Mumbai", "Bhopal", 3)
# g.add_edge("Agra", "Delhi", 1)

# g.print_adj_matrix()
# g.dijkstra("Delhi")

File: 33-graph-algorithms/14-dijkstra-algorithm-shortest-path.cpp
#include<bits/stdc++.h>
using namespace std;

template <typename T>
class Graph{
    unordered_map<T, list<pair<T, int> > > m; // {Delhi->[(Mumbai, 2), (Pune, 4)]}
    public:
    void addEdge(T u, T v, int dist, bool bidir=true){
        m[u].push_back(make_pair(v,dist));
        if(bidir)
            m[v].push_back(make_pair(u,dist));
    }
    void printAdj(){
        for(auto j:m){ // for map
            cout<<j.first<<"->";
            for(auto l: j.second){ // for list
                cout<<"("<<l.first<<","<<l.second<<")";
            }
            cout<<endl;
        }
    }
    void dijsktraSSSP(T src){
        unordered_map<T, int> dist;
        for(auto j: m){
            dist[j.first] = INT_MAX; // initialize each node INF distance
        }
        set<pair<int, T> > s; // first is distance, as set sort by first parameter
        dist[src] = 0; // source distance from itself is always 0
        s.insert(make_pair(0, src));

        while(!s.empty()){
            auto p = *(s.begin());
            T node = p.second;
            int nodeDist = p.first;
            s.erase(s.begin());

            for(auto childPair: m[node]){
                if(nodeDist + childPair.second < dist[childPair.first]){
                    T dest = childPair.first;
                    auto f = s.find( make_pair(dist[dest], dest) );
                    if(f!=s.end()){ // if destination in set , to update distance in set, remove then add with updated distance
                        s.erase(f);
                    }
                    dist[dest] = nodeDist + childPair.second;
                    s.insert(make_pair(dist[dest], dest));                }
            }
        }
        
        for(auto d:dist){
            cout<<d.first<<" is located at: "<<d.second<<endl;
        }

    }



};

int main() {
    // Graph<int> g;
    // g.addEdge(1,2,1);
    // g.addEdge(1,3,4);
    // g.addEdge(2,3,1);
    // g.addEdge(3,4,2);
    // g.printAdj();
    // g.dijsktraSSSP(1);

    Graph<string> india;
    india.addEdge("Amritsar", "Delhi", 1);
    india.addEdge("Amritsar", "Jaipur", 4);
    india.addEdge("Jaipur", "Delhi", 2);
    india.addEdge("Jaipur", "Mumbai", 8);
    india.addEdge("Bhopal", "Agra", 2);
    india.addEdge("Mumbai", "Bhopal", 3);
    india.addEdge("Agra", "Delhi", 1);
    india.printAdj();
    india.dijsktraSSSP("Amritsar");

    return 0;
}

/*
Single source shortest path algorithm.
From single source calculate shortest path to all the nodes.
Edges can be bi or uni-directional edges.
It uses weighted graph.
Adjacency list is most efficient way.


Algo:
    -> all nodes have INF distance except source.
    -> use priority queue(hashmap) or set.
    -> distance of node = distance of parent + weight (min).
    -> pick node with min distance from the set<node, distance>.
    -> set is sorted accoring to distance.
    -> to update in set -> remove then insert new pair.

Complexity: O(ElogV)
*/
